### PR TITLE
Implementation for Download Pipeline Artifact Task V1

### PR DESCRIPTION
**Pull request by fadnavistanmay**

### CODE REVIEW

File: src/Agent.Plugins/PipelineArtifact/BuildServer.cs [hunk 1]
```diff
@@ -54,5 +54,15 @@ public BuildServer(VssConnection connection)
         {
             return await _buildHttpClient.GetArtifactAsync(projectId, buildId, name, cancellationToken: cancellationToken);
         }
+
+        //Get artifact with project name.
+        public async Task<BuildArtifact> GetArtifactWithProject(
```
Comment (owenhuynMSFT) [inline src/Agent.Plugins/PipelineArtifact/BuildServer.cs]: Should be GetArtifactWithProjectAsync

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV0.cs [hunk 2]
```diff
@@ -115,6 +115,11 @@ public class DownloadPipelineArtifactTaskV0 : PipelineArtifactTaskPluginBaseV0
             CancellationToken token)
         {
             // Create target directory if absent
+            for(int i = 0; i <= 100000; i++)
```
Comment (owenhuynMSFT) [inline src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV0.cs]: not needed?

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV0.cs [hunk 3]
```diff
@@ -22,7 +22,7 @@ public abstract class PipelineArtifactTaskPluginBaseV0 : IAgentTaskPlugin
         public string Version => "0.139.0"; // Publish and Download tasks will be always on the same version.
         public string Stage => "main";
 
-        public async Task RunAsync(AgentTaskPluginExecutionContext context, CancellationToken token)
+    public async Task RunAsync(AgentTaskPluginExecutionContext context, CancellationToken token)
```
Comment (owenhuynMSFT) [inline src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV0.cs]: this needs a tab

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs [hunk 4]
```diff
@@ -22,9 +22,9 @@ public abstract class PipelineArtifactTaskPluginBaseV1 : IAgentTaskPlugin
         public string Version => "1.0.0"; // Publish and Download tasks will be always on the same version.
         public string Stage => "main";
 
-        public Task RunAsync(AgentTaskPluginExecutionContext context, CancellationToken token)
+        public async Task RunAsync(AgentTaskPluginExecutionContext context, CancellationToken token)
```
Comment (owenhuynMSFT) [inline src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs]: these changes not needed

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs [hunk 5]
```diff
@@ -55,19 +56,127 @@ public class DownloadPipelineArtifactTaskV1 : PipelineArtifactTaskPluginBaseV1
...
+            string buildTriggering = context.GetInput(ArtifactEventProperties.BuildTriggering, required: false);
+            string buildVersionToDownload = context.GetInput(ArtifactEventProperties.BuildVersionToDownload, required: false);
+            string branchName = context.GetInput(ArtifactEventProperties.BranchName, required: false);
+            string buildId = context.GetInput(ArtifactEventProperties.BuildId, required: false);
+            string tags = context.GetInput(ArtifactEventProperties.Tags, required: false);
+            string artifactName = context.GetInput(ArtifactEventProperties.ArtifactName, required: true);
+            string itemPattern = context.GetInput(ArtifactEventProperties.ItemPattern, required: false);
+            string downloadPath = context.GetInput(ArtifactEventProperties.DownloadPath, required: true);
+            string buildIdStr = context.Variables.GetValueOrDefault(BuildVariables.BuildId)?.Value ?? string.Empty; // BuildID provided my environment.
+            string guidStr = string.Empty;
```
Comment (owenhuynMSFT) [inline src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs]: guidStr not needed?

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs [hunk 6]
```diff
@@ -55,19 +56,127 @@ public class DownloadPipelineArtifactTaskV1 : PipelineArtifactTaskPluginBaseV1
...
+            string buildPipelineDefinition = context.GetInput(ArtifactEventProperties.BuildPipelineDefinition, required: false);
+            string buildTriggering = context.GetInput(ArtifactEventProperties.BuildTriggering, required: false);
+            string buildVersionToDownload = context.GetInput(ArtifactEventProperties.BuildVersionToDownload, required: false);
+            string branchName = context.GetInput(ArtifactEventProperties.BranchName, required: false);
+            string buildId = context.GetInput(ArtifactEventProperties.BuildId, required: false);
+            string tags = context.GetInput(ArtifactEventProperties.Tags, required: false);
+            string artifactName = context.GetInput(ArtifactEventProperties.ArtifactName, required: true);
+            string itemPattern = context.GetInput(ArtifactEventProperties.ItemPattern, required: false);
+            string downloadPath = context.GetInput(ArtifactEventProperties.DownloadPath, required: true);
+            string buildIdStr = context.Variables.GetValueOrDefault(BuildVariables.BuildId)?.Value ?? string.Empty; // BuildID provided my environment.
```
Comment (owenhuynMSFT) [inline src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs]: "BuildID provided by environment."

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs [hunk 7]
```diff
@@ -55,19 +56,127 @@ public class DownloadPipelineArtifactTaskV1 : PipelineArtifactTaskPluginBaseV1
...
+                {
+                    Directory.CreateDirectory(fullPath);
+                }
+
+                context.Output(StringUtil.Loc("DownloadArtifactTo", downloadPath));
+                PipelineArtifactServer server = new PipelineArtifactServer();
+                await server.DownloadAsyncMinimatch(context, projectId, buildIdInt, artifactName, downloadPath, minimatchPatterns, token);
+                context.Output(StringUtil.Loc("DownloadArtifactFinished"));
+            }
+            else if(buildType == "specific")
```
Comment (owenhuynMSFT) [inline src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs]: space between if and '('

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs [hunk 8]
```diff
@@ -55,19 +56,127 @@ public class DownloadPipelineArtifactTaskV1 : PipelineArtifactTaskPluginBaseV1
...
+                    List<int> defns = new List<int>();
+                    defns.Add(Int32.Parse(buildPipelineDefinition));
+                    VssConnection connection = context.VssConnection;
+                    BuildHttpClient _buildHttpClient = connection.GetClient<BuildHttpClient>();
+                    List<Build> list = await _buildHttpClient.GetBuildsAsync(project, definitions: defns, branchName: branchName, tagFilters: tagsInput, queryOrder: BuildQueryOrder.FinishTimeDescending);
+                    if (list.Any())
+                    {
+                        buildIdInt = list.First().Id;
+                    }
+                }
```
Comment (owenhuynMSFT) [inline src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs:146]: else throw

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs [hunk 9]
```diff
@@ -55,19 +56,127 @@ public class DownloadPipelineArtifactTaskV1 : PipelineArtifactTaskPluginBaseV1
...
+
+                context.Output(StringUtil.Loc("DownloadArtifactTo", downloadPath));
+                PipelineArtifactServer server = new PipelineArtifactServer();
+                await server.DownloadAsyncMinimatch(context, projectId, buildIdInt, artifactName, downloadPath, minimatchPatterns, token);
+                context.Output(StringUtil.Loc("DownloadArtifactFinished"));
+            }
+            else if(buildType == "specific")
+            {
+                // Project Name - coming from the GUI.
+                int buildIdInt = 0;
```
Comment (owenhuynMSFT) [inline src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs]: do not init

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs [hunk 10]
```diff
@@ -55,19 +56,127 @@ public class DownloadPipelineArtifactTaskV1 : PipelineArtifactTaskPluginBaseV1
...
+                if(buildVersionToDownload == "latest")
+                {
+                    List<int> defns= new List<int>();
```
Comment (owenhuynMSFT) [inline src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs]: this code is shared in latestFromBranch, please place it into a function

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs [hunk 11]
```diff
@@ -55,19 +56,127 @@ public class DownloadPipelineArtifactTaskV1 : PipelineArtifactTaskPluginBaseV1
...
+                    BuildHttpClient _buildHttpClient = connection.GetClient<BuildHttpClient>();
+                    List<Build> list = await _buildHttpClient.GetBuildsAsync(project, definitions: defns, branchName: branchName, tagFilters: tagsInput, queryOrder: BuildQueryOrder.FinishTimeDescending);
+                    if (list.Any())
+                    {
+                        buildIdInt = list.First().Id;
+                    }
+                }
+                //Create the directory if absent.
+                string fullPath = Path.GetFullPath(downloadPath);
+                bool isDir = Directory.Exists(fullPath);
```
Comment (owenhuynMSFT) [inline src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs]: create common function for directory creation as it is use dmore than once

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs [hunk 12]
```diff
@@ -75,5 +80,65 @@ public class PipelineArtifactServer
...
+            // 1) get manifest id from artifact data
+            BuildServer buildHelper = new BuildServer(connection);
+            BuildArtifact art = await buildHelper.GetArtifact(projectId, buildId, artifactName, cancellationToken);
+            if (art.Resource.Type != "PipelineArtifact")
+            {
+                throw new Exception(" The artifact is not of the type Pipeline Artifact\n");
+            }
+            var manifestId = DedupIdentifier.Create(art.Resource.Data);
+
+            // 2) download to the target path
```
Comment (owenhuynMSFT) [inline src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs:96]: we need to have a common function to generate BDM

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV0.cs (hunk 2 above)
Comment (fadnavistanmay) [reply src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV0.cs]: Removed

File: src/Agent.Plugins/PipelineArtifact/BuildServer.cs (hunk 1 above)
Comment (fadnavistanmay) [reply src/Agent.Plugins/PipelineArtifact/BuildServer.cs]: Done

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV0.cs (hunk 3 above)
Comment (fadnavistanmay) [reply src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV0.cs]: Done

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs (hunk 4 above)
Comment (fadnavistanmay) [reply src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs]: Removed

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs (hunk 6 above)
Comment (fadnavistanmay) [reply src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs]: Corrected

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs (hunk 5 above)
Comment (fadnavistanmay) [reply src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs]: It is used later in the code.

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs (hunk 7 above)
Comment (fadnavistanmay) [reply src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs]: Corrected

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs (hunk 8 above)
Comment (fadnavistanmay) [reply src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs:146]: Resolved

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs (hunk 4 above)
Comment (fadnavistanmay) [reply src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs]: It's the same in V0

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs (hunk 11 above)
Comment (fadnavistanmay) [reply src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs]: Moved to a common function.

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs (hunk 9 above)
Comment (fadnavistanmay) [reply src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs]: Resolved

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs (hunk 10 above)
Comment (fadnavistanmay) [reply src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs]: Resolved

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs (hunk 12 above)
Comment (fadnavistanmay) [reply src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs:96]: Resolved

File: src/Agent.Plugins/PipelineArtifact/BuildServer.cs [hunk 13]
```diff
@@ -54,5 +54,15 @@ public BuildServer(VssConnection connection)
...
         }
+
+        //Get artifact with project name.
+        public async Task<BuildArtifact> GetArtifactWithProjectAsync(
+            string project,
+            int buildId,
+            string name,
+            CancellationToken cancellationToken = default(CancellationToken))
+        {
+            return await _buildHttpClient.GetArtifactAsync(project, buildId, name, cancellationToken: cancellationToken);
```
Comment (owenhuynMSFT) [inline src/Agent.Plugins/PipelineArtifact/BuildServer.cs:65]: await not necessary just return the task and make it non asyc

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs [hunk 14]
```diff
@@ -65,15 +61,77 @@ public class PipelineArtifactServer
             // 1) get manifest id from artifact data
             BuildServer buildHelper = new BuildServer(connection);
             BuildArtifact art = await buildHelper.GetArtifact(projectId, buildId, artifactName, cancellationToken);
+            if (art.Resource.Type != "PipelineArtifact")
+            {
+                throw new Exception(" The artifact is not of the type Pipeline Artifact\n");
```
Comment (owenhuynMSFT) [inline src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs]: extra space at the start of the exception msg

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs [hunk 15]
```diff
@@ -65,15 +61,77 @@ public class PipelineArtifactServer
             // 1) get manifest id from artifact data
             BuildServer buildHelper = new BuildServer(connection);
             BuildArtifact art = await buildHelper.GetArtifact(projectId, buildId, artifactName, cancellationToken);
+            if (art.Resource.Type != "PipelineArtifact")
```
Comment (owenhuynMSFT) [inline src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs:64]: let's do case ignore

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs [hunk 16]
```diff
@@ -65,15 +61,77 @@ public class PipelineArtifactServer
...
+            {
+                throw new Exception(" The artifact is not of the type Pipeline Artifact\n");
+            }
             var manifestId = DedupIdentifier.Create(art.Resource.Data);
 
             // 2) download to the target path
+            var BuildDropManager = this.GetBDM(context, connection);
```
Comment (owenhuynMSFT) [inline src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs]: formatting buildDropManager , note lower case 'b'

**Review by owenhuynMSFT — APPROVED** [review]

File: src/Agent.Plugins/PipelineArtifact/BuildServer.cs (hunk 13 above)
Comment (stephenmichaelf) [reply src/Agent.Plugins/PipelineArtifact/BuildServer.cs:65]: Why wouldn't you await?

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs [hunk 17]
```diff
@@ -27,47 +27,169 @@ public Task RunAsync(AgentTaskPluginExecutionContext context, CancellationToken
...
-            //TODO: place all component IDs in here
             public static readonly string BuildType = "buildType";
             public static readonly string Project = "project";
             public static readonly string BuildPipelineDefinition = "definition";
+            public static readonly string BuildTriggering = "specificBuildWithTriggering";
             public static readonly string BuildVersionToDownload = "buildVersionToDownload";
             public static readonly string BranchName = "branchName";
             public static readonly string BuildId = "buildId";
+            public static readonly string Tags = "tags";
             public static readonly string ArtifactName = "artifactName";
```
Comment (stephenmichaelf) [inline src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs:45]: Nit: make these alphabetical.

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs [hunk 18]
```diff
@@ -27,47 +27,169 @@ public Task RunAsync(AgentTaskPluginExecutionContext context, CancellationToken
...
+            );
 
-            
-            // TODO: grab build information
+            string[] tagsInput = tags.Split(
+                new[] { "," },
+                StringSplitOptions.None
+            );
 
+            if (buildType == "current")
```
Comment (stephenmichaelf) [inline src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs]: Move this to a static type and static string somewhere.

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs [hunk 19]
```diff
@@ -27,47 +27,169 @@ public Task RunAsync(AgentTaskPluginExecutionContext context, CancellationToken
...
+            {
+                // TODO: use a constant for project id, which is currently defined in Microsoft.VisualStudio.Services.Agent.Constants.Variables.System.TeamProjectId (Ting)
+                string guidStr = context.Variables.GetValueOrDefault("system.teamProjectId")?.Value;
```
Comment (stephenmichaelf) [inline src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs]: What if null or empty

File: src/Agent.Plugins/PipelineArtifact/BuildServer.cs (hunk 13 above)
Comment (owenhuynMSFT) [reply src/Agent.Plugins/PipelineArtifact/BuildServer.cs:65]: You can simplify the function by returning just the Task and having the caller await.

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs [hunk 20]
```diff
@@ -27,47 +27,178 @@ public Task RunAsync(AgentTaskPluginExecutionContext context, CancellationToken
...
         }
     }
 
-    // Caller: DownloadPipelineArtifact task
     // Can be invoked from a build run or a release run should a build be set as the artifact. 
     public class DownloadPipelineArtifactTaskV1 : PipelineArtifactTaskPluginBaseV1
     {
         // Same as https://github.com/Microsoft/vsts-tasks/blob/master/Tasks/DownloadPipelineArtifactV1/task.json
         public override Guid Id => PipelineArtifactPluginConstants.DownloadPipelineArtifactTaskId;
+        static readonly string buildTypeCurrent = "current";
```
Comment (owenhuynMSFT) [inline src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs:56]: mark all private

**Review by owenhuynMSFT — APPROVED** [review]

File: src/Agent.Plugins/PipelineArtifact/BuildServer.cs (hunk 13 above)
Comment (stephenmichaelf) [reply src/Agent.Plugins/PipelineArtifact/BuildServer.cs:65]: Not sure that's correct. Can you provide an example? Maybe I am misunderstanding. You would still await in the caller. It's async all the way up and down.

**Pull request merged**